hoomd/md/nlist.py

```diff
@@ -104,10 +115,65 @@ def __init__(self, buffer, exclusions, rebuild_check_delay, check_dist,
 
         self._mesh = validate_mesh(mesh)
 
+        self._in_context_manager = False
+
     def _attach_hook(self):
         if self._mesh is not None:
             self._cpp_obj.addMesh(self._mesh._cpp_obj)
 
+        super()._attach()
+
```

Comment:
Don't call this directly (actually this is calling `_attach_hook`).

Reply:
Yes I noticed that after I pushed. I applied it last night, but it didn't have an effect.

Reply:
So `_attach` is responsible in keeping track of the number of places an object is in a simulation. `_attach_hook` is for creating the C++ object. All consumers of `_HOOMDBaseObject` subclasses must use `_attach` to increment the object counter. This has no other effect if already attached.

Detaching is similar `_detach` handles decrementing the counter and when reaching 0 handles object clean up. Most classes have no need of a `_detach_hook` method, but the method is provided to allow for special detaching behavior. Consumers are likewise obligated to call `_detach` when removing objects or being removed themselves.

Reply:
Issue appears to be stemming from these lines, which prematurely dissociates the Simulation from the nlist. https://github.com/glotzerlab/hoomd-blue/blob/289939f469e0f866ada8f257331409d6be1d3bbb/hoomd/data/syncedlist.py#L212-L213

This seems to be the right logic?

``` python
if not value._attached and value._added:
    value._remove()
```

Reply:
Seems correct and the new logic seems right. Problem here is that adding and attaching are different steps, and there is only one counter, but checking the `_attach` property should fix this.

`added` and `removed` may now actually be unnecessary as they mostly served to ensure that the same object was not added to multiple lists, but we cannot do this anymore given that the objects may be added to consumers that themselves are associated with a simulation. I will think on this.

Reply:
This doesn't appear to be the only case either. Removing a force while an nlist is attached as a compute causes the counter to be decremented twice.

``` python
nlist = hoomd.md.nlist.Tree(buffer=0, default_r_cut=1.01)
sim: hoomd.Simulation = simulation_factory(lattice_snapshot_factory())

integrator = hoomd.md.Integrator(0.005)
lj = hoomd.md.pair.LJ(nlist, default_r_cut=1.1)
lj.params[('A', 'A')] = dict(epsilon=1.0, sigma=1.0)
integrator.forces.append(lj)
integrator.methods.append(hoomd.md.methods.NVE(hoomd.filter.All()))
sim.operations.integrator = integrator

sim.operations.computes.append(nlist)
print(len(sim.operations.computes))  # 1
print(nlist._use_count)  # 2
sim.operations.integrator.forces.clear()
print(nlist._use_count)  # 0
print(len(sim.operations.computes))  # 1
```

Reply:
Yeah, I imagine this is happening due to the way the `Pair` class is implemented. Sorry for the bugs the logic was fairly complicated when we were restricting objects to one "container", so changing it over is difficult. I am creating a PR that should fix these errors by removing `_add()` and `_remove()`.

Reply:
No worries! Yea I understand, let me know if I can help at all with any of these fixes!

Reply:
PR #1418 should fix this.